File: sysplant/templates/iterators/canterlot.c
// Parse Exception Directory and lookup syscall by name (sort by ssn number) https://www.mdsec.co.uk/2022/04/resolving-system-service-numbers-using-the-exception-directory/
BOOL SPT_PopulateSyscallList(void)
{
    // Return early if the list is already populated.
    if (SPT_SyscallList.Count) return TRUE;

#if defined(_WIN64)
    // If the process is 64-bit on a 64-bit OS, we need to search for syscall
    PSPT_PEB Peb = (PSPT_PEB)__readgsqword(0x60);
#else
    // If the process is 32-bit on a 32-bit OS, we need to search for sysenter
    PSPT_PEB Peb = (PSPT_PEB)__readfsdword(0x30);
#endif
    PSPT_PEB_LDR_DATA Ldr = Peb->Ldr;
    PIMAGE_EXPORT_DIRECTORY ExportDirectory = NULL;
    PIMAGE_RUNTIME_FUNCTION_ENTRY RuntimeFunctions = NULL;
    PVOID DllBase = NULL;
    

    // Get the DllBase address of NTDLL.dll. NTDLL is not guaranteed to be the second
    // in the list, so it's safer to loop through the full list and find it.
    PSPT_LDR_DATA_TABLE_ENTRY LdrEntry;
    for (LdrEntry = (PSPT_LDR_DATA_TABLE_ENTRY)Ldr->Reserved2[1]; LdrEntry->DllBase != NULL; LdrEntry = (PSPT_LDR_DATA_TABLE_ENTRY)LdrEntry->Reserved1[0])
    {
        DllBase = LdrEntry->DllBase;
        PIMAGE_DOS_HEADER DosHeader = (PIMAGE_DOS_HEADER)DllBase;
        PIMAGE_NT_HEADERS NtHeaders = SPT_RVA2VA(PIMAGE_NT_HEADERS, DllBase, DosHeader->e_lfanew);
        PIMAGE_DATA_DIRECTORY DataDirectory = (PIMAGE_DATA_DIRECTORY)NtHeaders->OptionalHeader.DataDirectory;
        DWORD ExportVirtualAddress = DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].VirtualAddress;
        if (ExportVirtualAddress == 0) continue; // no export table? skip

        // Load ExportDirectory
        ExportDirectory = (PIMAGE_EXPORT_DIRECTORY)SPT_RVA2VA(ULONG_PTR, DllBase, ExportVirtualAddress);
        // Load the Exception Directory.
        DWORD ExceptionVirtualAddress = DataDirectory[IMAGE_DIRECTORY_ENTRY_EXCEPTION].VirtualAddress;
        RuntimeFunctions = SPT_RVA2VA(PIMAGE_RUNTIME_FUNCTION_ENTRY, DllBase, ExceptionVirtualAddress);

        // If this is NTDLL.dll, exit loop.
        PCHAR DllName = SPT_RVA2VA(PCHAR, DllBase, ExportDirectory->Name);

        // Ntdll.dll check corrected to remove compiler warning
        if ((*(ULONG*)DllName | 0x20202020) != 0x6c64746e) continue;
        if ((*(ULONG*)(DllName + 4) | 0x20202020) == 0x6c642e6c) break;
    }

    if (!RuntimeFunctions) return FALSE;

    // Load the Export Address Table
    PDWORD Functions = SPT_RVA2VA(PDWORD, DllBase, ExportDirectory->AddressOfFunctions);
    PDWORD Names = SPT_RVA2VA(PDWORD, DllBase, ExportDirectory->AddressOfNames);
    PWORD Ordinals = SPT_RVA2VA(PWORD, DllBase, ExportDirectory->AddressOfNameOrdinals);

    // Create Syscall list Entry
    PSPT_SYSCALL_ENTRY Entries = SPT_SyscallList.Entries;

    WORD ssn = 0;
    WORD padding = 0x0;
    // Search runtime function table.
    for (DWORD i = 0; RuntimeFunctions[i].BeginAddress; i++) {
        // Search export address table.
        for (DWORD cx = 0; cx < ExportDirectory->NumberOfNames; cx++) {
            PCHAR FunctionName = SPT_RVA2VA(PCHAR, DllBase, Names[cx]);
            PVOID FunctionAddress = SPT_RVA2VA(PVOID, DllBase, Functions[Ordinals[cx]]);

            // begin address rva?
            if (Functions[Ordinals[cx]] == RuntimeFunctions[i].BeginAddress) {
                // Is this a system call?
                if (*(USHORT*)FunctionName == 0x775a)
                {
                    Entries[ssn].Hash = SPT_HashSyscallName(FunctionName);
                    Entries[ssn].Address = FunctionAddress;

                    // All syscall stubs are identical for a Windows version
                    if (padding == 0x0) {
                        padding = SPT_DetectPadding(FunctionAddress);
                    }

                    // Set syscall entry with appropriate padding
                    Entries[ssn].SyscallAddress = SPT_RVA2VA(PVOID, Entries[ssn].Address, padding);
                    
                    // Increase the ssn value
                    ssn++;

                    // Update number of system calls found.
                    SPT_SyscallList.Count = ssn;

                    // No need to go further for this one
                    break;
                }
            }
        }
    }

    return TRUE;
}

File: sysplant/templates/iterators/canterlot.rs
// Parse Exception Directory and lookup syscall by name (sort by ssn number)
// https://www.mdsec.co.uk/2022/04/resolving-system-service-numbers-using-the-exception-directory/
#[unsafe(no_mangle)]
unsafe fn SPT_PopulateSyscallList() -> BOOL {
    // Return early if the list is already populated.
    if SPT_SyscallList.Count != 0 {
        return TRUE;
    }

    let peb = get_peb();
    let ldr = (*peb).Ldr;
    let mut export_directory: *mut IMAGE_EXPORT_DIRECTORY = ptr::null_mut();
    let mut runtime_functions: *mut IMAGE_RUNTIME_FUNCTION_ENTRY = ptr::null_mut();
    let mut dll_base: PVOID = ptr::null_mut();

    // Get the DllBase address of NTDLL.dll. NTDLL is not guaranteed to be the second
    // in the list, so it's safer to loop through the full list and find it.
    let mut ldr_entry = (*ldr).Reserved2[1] as *mut SPT_LDR_DATA_TABLE_ENTRY;
    while !(*ldr_entry).DllBase.is_null() {
        dll_base = (*ldr_entry).DllBase;
        let dos_header = dll_base as *const IMAGE_DOS_HEADER;
        let nt_headers = SPT_RVA2VA!(*const IMAGE_NT_HEADERS64, dll_base, (*dos_header).e_lfanew);
        let data_directory = (*nt_headers).OptionalHeader.DataDirectory.as_ptr();
        let export_va = (*data_directory.add(IMAGE_DIRECTORY_ENTRY_EXPORT)).VirtualAddress;
        if export_va == 0 {
            ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }

        // Load ExportDirectory
        export_directory = SPT_RVA2VA!(*mut IMAGE_EXPORT_DIRECTORY, dll_base, export_va);
        // Load the Exception Directory.
        let exception_va = (*data_directory.add(IMAGE_DIRECTORY_ENTRY_EXCEPTION)).VirtualAddress;
        runtime_functions = SPT_RVA2VA!(*mut IMAGE_RUNTIME_FUNCTION_ENTRY, dll_base, exception_va);

        // If this is NTDLL.dll, exit loop.
        let dll_name = SPT_RVA2VA!(*const u8, dll_base, (*export_directory).Name);
        // "ntdl" case-insensitive check
        if (*(dll_name as *const u32) | 0x20202020) != 0x6c64746e {
            ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }
        // "l.dl" case-insensitive check
        if (*(dll_name.add(4) as *const u32) | 0x20202020) == 0x6c642e6c {
            break;
        }
        ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
    }

    if runtime_functions.is_null() {
        return FALSE;
    }

    // Load the Export Address Table
    let functions = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfFunctions);
    let names = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfNames);
    let ordinals = SPT_RVA2VA!(*const WORD, dll_base, (*export_directory).AddressOfNameOrdinals);

    let entries = SPT_SyscallList.Entries.as_mut_ptr();
    let mut ssn: u16 = 0;
    let mut padding: u16 = 0;

    // Search runtime function table.
    let mut i: u32 = 0;
    while (*runtime_functions.add(i as usize)).BeginAddress != 0 {
        // Search export address table.
        for cx in 0..(*export_directory).NumberOfNames {
            let function_name = SPT_RVA2VA!(*const i8, dll_base, *names.add(cx as usize));
            let function_address = SPT_RVA2VA!(PVOID, dll_base, *functions.add(*ordinals.add(cx as usize) as usize));

            // begin address rva?
            if *functions.add(*ordinals.add(cx as usize) as usize) == (*runtime_functions.add(i as usize)).BeginAddress {
                // Is this a system call? "Zw" == 0x775a
                if *(function_name as *const u16) == 0x775a {
                    (*entries.add(ssn as usize)).Hash = SPT_HashSyscallName(function_name);
                    (*entries.add(ssn as usize)).Address = function_address;

                    // All syscall stubs are identical for a Windows version
                    if padding == 0 {
                        padding = SPT_DetectPadding(function_address);
                    }

                    // Set syscall entry with appropriate padding
                    (*entries.add(ssn as usize)).SyscallAddress = SPT_RVA2VA!(PVOID, (*entries.add(ssn as usize)).Address, padding);

                    ssn += 1;
                    SPT_SyscallList.Count = ssn as DWORD;
                    break;
                }
            }
        }
        i += 1;
    }

    TRUE
}
